File: src/main/java/work/lclpnet/ap2/impl/game/data/EliminationDataContainer.java
package work.lclpnet.ap2.impl.game.data;

import it.unimi.dsi.fastutil.objects.Object2IntArrayMap;
import it.unimi.dsi.fastutil.objects.Object2IntMap;
import work.lclpnet.ap2.api.game.data.DataContainer;
import work.lclpnet.ap2.api.game.data.DataEntry;
import work.lclpnet.ap2.api.game.data.SubjectRef;
import work.lclpnet.ap2.api.game.data.SubjectRefFactory;
import work.lclpnet.ap2.impl.game.data.entry.SimpleOrderDataEntry;
import work.lclpnet.kibu.translate.text.TranslatedText;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Stream;

/**
 * A data container for last one standing game modes.
 * The last added subject is the winner.
 */
public class EliminationDataContainer<T, Ref extends SubjectRef> implements DataContainer<T, Ref> {

    private final SubjectRefFactory<T, Ref> refs;
    private final Object2IntMap<Ref> index = new Object2IntArrayMap<>();
    private final List<Map<Ref, SimpleOrderDataEntry<Ref>>> order = new ArrayList<>();
    private boolean frozen = false;

    public EliminationDataContainer(SubjectRefFactory<T, Ref> refs) {
        this.refs = refs;
    }

    public void eliminated(T subject) {
        eliminated(subject, null);
    }

    public void eliminated(T subject, TranslatedText data) {
        allEliminated(List.of(subject), data);
    }

    public void allEliminated(Iterable<? extends T> subjects) {
        allEliminated(subjects, null);
    }

    public void allEliminated(Iterable<? extends T> subjects, TranslatedText data) {
        synchronized (this) {
            if (frozen) return;

            int rank = order.size();
            Map<Ref, SimpleOrderDataEntry<Ref>> mapping = new HashMap<>();

            for (T subject : subjects) {
                Ref ref = refs.create(subject);

                if (index.containsKey(ref)) continue;

                index.put(ref, rank);
                mapping.put(ref, new SimpleOrderDataEntry<>(ref, rank, data));
            }

            if (mapping.isEmpty()) return;

            order.add(mapping);
        }
    }

    @Override
    public void delete(T subject) {
        synchronized (this) {
            if (frozen) return;

            Ref ref = refs.create(subject);
            int rank = getRank(ref);

            if (rank < 0 || rank >= order.size()) return;

            index.removeInt(ref);

            var entries = order.get(rank);

            if (entries == null) return;

            entries.remove(ref);
        }
    }

    @Override
    public DataEntry<Ref> getEntry(T subject) {
        Ref ref = refs.create(subject);

        synchronized (this) {
            int rank = getRank(ref);

            if (rank < 0 || rank >= order.size()) {
                return new SimpleOrderDataEntry<>(ref, -1);
            }

            var mapping = order.get(rank);

            if (mapping == null) {
                return new SimpleOrderDataEntry<>(ref, rank);
            }

            var dataEntry = mapping.get(ref);

            if (dataEntry == null) {
                return new SimpleOrderDataEntry<>(ref, rank);
            }

            return dataEntry;
        }
    }

    @Override
    public Stream<DataEntry<Ref>> orderedEntries() {
        synchronized (this) {
            return this.order.stream().flatMap(mapping -> mapping.values().stream());
        }
    }

    @Override
    public void freeze() {
        synchronized (this) {
            frozen = true;
        }
    }

    @Override
    public void ensureTracked(T subject) {
        eliminated(subject);
    }

    private int getRank(Ref ref) {
        return index.getOrDefault(ref, -1);
    }
}

File: src/lib/src/main/kotlin/work/lclpnet/ap2/game/data/EliminationDataContainer.kt
package work.lclpnet.ap2.game.data

import it.unimi.dsi.fastutil.objects.Object2IntArrayMap
import work.lclpnet.ap2.api.game.data.DataContainer
import work.lclpnet.ap2.api.game.data.DataEntry
import work.lclpnet.ap2.api.game.data.SubjectRef
import work.lclpnet.ap2.api.game.data.SubjectRefFactory
import work.lclpnet.ap2.game.data.entry.SimpleOrderDataEntry
import work.lclpnet.kibu.translate.text.TranslatedText
import java.util.stream.IntStream
import java.util.stream.Stream

/**
 * A data container for last one standing game modes.
 * The last added subject is the winner.
 */
class EliminationDataContainer<T, Ref : SubjectRef>(
    refs: SubjectRefFactory<T, Ref>
) : BaseDataContainer<T, Ref>(refs) {

    private val index = Object2IntArrayMap<Ref>()
    private val order = ArrayList<MutableMap<Ref, SimpleOrderDataEntry<Ref>>>()

    override fun add(subject: T) {
        add(subject, null)
    }

    fun add(subject: T, data: TranslatedText?) {
        addAll(listOf(subject), data)
    }

    fun addAll(subjects: Iterable<T>) {
        addAll(subjects, null)
    }

    @Synchronized
    fun addAll(subjects: Iterable<T>, data: TranslatedText?) {
        val rank = order.size
        val mapping = HashMap<Ref, SimpleOrderDataEntry<Ref>>()

        for (subject in subjects) {
            val ref = refs.create(subject)

            if (index.containsKey(ref)) continue

            index.put(ref, rank)
            mapping[ref] = SimpleOrderDataEntry(ref, rank, data)
        }

        if (mapping.isEmpty()) return

        order.add(mapping)
    }

    @Synchronized
    override fun getEntry(ref: Ref): DataEntry<Ref>? {
        val rank = getRank(ref)

        if (rank < 0 || rank >= order.size) {
            return null
        }

        val mapping = order[rank]
        val dataEntry = mapping[ref]

        return dataEntry
    }

    @Synchronized
    override fun streamOrderedEntries(): Stream<DataEntry<Ref>> {
        // order needs to be reversed
        return IntStream.range(0, order.size)
            .mapToObj { i -> order[order.size - i - 1] }
            .flatMap { mapping -> mapping.values.stream() }
    }

    override fun identityIfAbsent(subject: T) {}

    @Synchronized
    override fun clear() {
        index.clear()
        order.clear()
    }

    private fun getRank(ref: Ref): Int {
        return index.getOrDefault(ref, -1)
    }

    @Synchronized
    override fun copy(): DataContainer<T, Ref> {
        val copy = EliminationDataContainer(refs)

        copy.index.putAll(this.index)
        copy.order.addAll(this.order)

        return copy
    }
}
